src/misc.c

```diff
@@ -159,23 +159,29 @@ _libssh2_recv(libssh2_socket_t sock, void *buffer, size_t length,
     (void)abstract;
 
     rc = recv(sock, buffer, length, flags);
+    if(rc < 0) {
+        int err;
 #ifdef WIN32
-    if(rc < 0)
-        return -wsa2errno();
+        err = _libssh2_wsa2errno();
 #else
-    if(rc < 0) {
+        err = errno;
+#endif
+        /* Profiling tools that use SIGPROF can cause EINTR responses.
+           recv() does not modify its arguments when it returns EINTR,
+           but there may be data waiting, so the caller should try again */
+        if(err == EINTR)
+            return 0;
```

Comment:
I observed in tests that returning `-EAGAIN` here also passes. That might be more consistent with this block of code.